qualtran/bloqs/chemistry/sparse/prepare.py

```diff
@@ -340,75 +345,48 @@ def build_qrom_bloq(self) -> 'Bloq':
-        beta: 'SoquetT',
-        rot_aa: 'SoquetT',
-        swap_pq: 'SoquetT',
-        swap_rs: 'SoquetT',
-        swap_pqrs: 'SoquetT',
-        flag_1b: 'SoquetT',
-        alt_pqrs: NDArray[Soquet],  # type: ignore[type-var]
-        theta: NDArray[Soquet],  # type: ignore[type-var]
-        keep: 'SoquetT',
-        less_than: 'SoquetT',
-        alt_flag_1b: 'SoquetT',
-    ) -> Dict[str, 'SoquetT']:
+    def add_qrom(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        qrom = self.build_qrom_bloq()
+        # The qroam_junk_regs won't be present initially when building the
+        # composite bloq as they're RIGHT registers.
+        qroam_out_soqs = bb.add_d(qrom, selection=soqs['d'])
+        out_soqs: Dict[str, 'SoquetT'] = {'d': qroam_out_soqs.pop('selection')}
+        # map output soqs to Prepare junk registers names
+        out_soqs |= {
+            reg.name: qroam_out_soqs.pop(f'target{i}_')
+            for (i, reg) in enumerate(self.qroam_junk_regs)
+        }
+        out_soqs |= {
+            reg.name: qroam_out_soqs.pop(f'junk_target{i}_')
+            for (i, reg) in enumerate(self.qroam_extra_junk_regs)
+        }
+        return soqs | out_soqs
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        n_n = self.num_bits_spat_orb
         # 1. Prepare \sum_d |d\rangle
-        d = bb.add(PrepareUniformSuperposition(self.num_non_zero), target=d)
+        soqs['d'] = bb.add(PrepareUniformSuperposition(self.num_non_zero), target=soqs['d'])
         # 2. QROM the ind_d alt_d values
-        qrom = self.build_qrom_bloq()
-        (
-            d,
-            p,
-            q,
-            r,
-            s,
-            theta[0],
-            flag_1b,
-            alt_pqrs[0],
-            alt_pqrs[1],
-            alt_pqrs[2],
-            alt_pqrs[3],
-            theta[1],
-            alt_flag_1b,
-            keep,
-        ) = bb.add(
-            qrom,
-            selection=d,
-            target0_=p,
-            target1_=q,
-            target2_=r,
-            target3_=s,
-            target4_=theta[0],
-            target5_=flag_1b,
-            target6_=alt_pqrs[0],
-            target7_=alt_pqrs[1],
-            target8_=alt_pqrs[2],
-            target9_=alt_pqrs[3],
-            target10_=theta[1],
-            target11_=alt_flag_1b,
-            target12_=keep,
-        )
-        lte_bloq = LessThanEqual(self.num_bits_state_prep, self.num_bits_state_prep)
+        soqs |= self.add_qrom(bb, **soqs)
         # prepare uniform superposition over sigma
-        sigma = bb.add(OnEach(self.num_bits_state_prep, Hadamard()), q=sigma)
-        keep, sigma, less_than = bb.add(lte_bloq, x=keep, y=sigma, target=less_than)
-        less_than, theta[1] = bb.add(CZ(), q1=less_than, q2=theta[1])
+        soqs['sigma'] = bb.add(OnEach(self.num_bits_state_prep, Hadamard()), q=soqs['sigma'])
+        # Inequality test for alias sampling
```

Comment:
Unrelated to the PR -- but do you know why are we implementing the alias sampling circuit explicitly here instead of directly relying on using `StatePreparationViaAliasSampling` bloq ? If it makes sense, Can you open an issue to document the requirements that would enable us to directly use the alias sampling bloq here?

Reply:
Well in this case we're using QROAM not QROM and our QROAM is quite special. Not sure if there is a more general abstraction to be made for SparseStatePreparationAliasSampling, but maybe?

Reply:
this has a single selection registers, multiple ind / alt registers, uses QROAM AND (not implemented here) uses swaps which can be inverted for free using MBUC (a la #1121 (I think))

Reply:
Ideally interchanging the QROMs for all the state prep bloqs would be nice